File: Bronze/2018/US Open/01_team_tic_tac_toe/TeamTicTacToe.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;

/**
 * @author Red
 * @version 1.0
 */
public class TeamTicTacToe {
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("tttt.in"));
        PrintWriter pw = new PrintWriter("tttt.out");
        char[][] board = new char[3][3];
        for (int i = 0; i < 3; i++) {
            board[i] = r.readLine().toCharArray();
        }
        int singleCount = 0;
        int teamCount = 0;
        for (char c1 = 'A'; c1 <= 'Z'; c1++) {
            // 单牛获胜
            if (singleWin(c1, board)) {
                singleCount++;
            }
            for (char c2 = 'Z'; c2 > c1; c2--) {
                // 双牛获胜  teamWin 中已实现 ZA 和 AZ 为相同组合
                if (teamWin(c1, c2, board)) {
                    teamCount++;
                }
            }
        }
        pw.println(singleCount);
        pw.println(teamCount);
        pw.close();
    }

    /**
     * 判断给定单牛是否获胜
     *
     * @param target 目标牛
     * @param board  棋盘
     * @return 是否获胜
     */
    static boolean singleWin(char target, char[][] board) {
        if (board[0][0] == target && board[1][1] == target && board[2][2] == target) {
            return true;
        }
        if (board[0][2] == target && board[1][1] == target && board[2][0] == target) {
            return true;
        }
        for (int i = 0; i < 3; i++) {
            if (board[i][0] == target && board[i][1] == target && board[i][2] == target) {
                return true;
            }
            if (board[0][i] == target && board[1][i] == target && board[2][i] == target) {
                return true;
            }
        }
        return false;
    }

    /**
     * 判断给定双牛是否获胜
     *
     * @param target1 牛1
     * @param target2 牛2
     * @param board   棋盘
     * @return 是否获胜
     */
    static boolean teamWin(char target1, char target2, char[][] board) {
        for (int i = 0; i < 3; i++) {
            if (match2(target1, target2, board[i][0], board[i][1], board[i][2])) {
                return true;
            }
            if (match2(target1, target2, board[0][i], board[1][i], board[2][i])) {
                return true;
            }
        }
        if (match2(target1, target2, board[0][0], board[1][1], board[2][2])) {
            return true;
        }
        if (match2(target1, target2, board[0][2], board[1][1], board[2][0])) {
            return true;
        }
        return false;
    }

    /**
     * 判断给定三个字符是否由两个目标字符组成
     * 即双牛的获胜逻辑
     */
    static boolean match2(char target1, char target2, char c1, char c2, char c3) {
        return (c1 == target1 && c2 == target2 && c3 == target2) ||
                (c1 == target2 && c2 == target1 && c3 == target2) ||
                (c1 == target2 && c2 == target2 && c3 == target1) ||
                (c1 == target2 && c2 == target1 && c3 == target1) ||
                (c1 == target1 && c2 == target2 && c3 == target1) ||
                (c1 == target1 && c2 == target1 && c3 == target2);
    }
}

File: Bronze/2018/US Open/01_team_tic_tac_toe/team_tic_tac_toe.py

fin = open('tttt.in', 'r')
fout = open("tttt.out", "w")
# 单牛获胜
def single_win(target, board):
    if board[0][0] == target and board[1][1] == target and board[2][2] == target:
        return True
    if board[0][2] == target and board[1][1] == target and board[2][0] == target:
        return True
    for i in range(3):
        if board[i][0] == target and board[i][1] == target and board[i][2] == target:
            return True
        if board[0][i] == target and board[1][i] == target and board[2][i] == target:
            return True
    return False
# 双牛获胜
def team_win(target1, target2, board):
    for i in range(3):
        if match2(target1, target2, board[i][0], board[i][1], board[i][2]):
            return True
        if match2(target1, target2, board[0][i], board[1][i], board[2][i]):
            return True
    if match2(target1, target2, board[0][0], board[1][1], board[2][2]):
        return True
    if match2(target1, target2, board[0][2], board[1][1], board[2][0]):
        return True
    return False
def match2(target1, target2, c1, c2, c3):
    return (c1 == target1 and c2 == target2 and c3 == target2) or \
        (c1 == target2 and c2 == target1 and c3 == target2) or \
        (c1 == target2 and c2 == target2 and c3 == target1) or \
        (c1 == target2 and c2 == target1 and c3 == target1) or \
        (c1 == target1 and c2 == target2 and c3 == target1) or \
        (c1 == target1 and c2 == target1 and c3 == target2)

board = [list(fin.readline().strip()) for _ in range(3)]
single_count = 0
team_count = 0

for c1 in range(ord('A'), ord('Z') + 1):
    # 单牛获胜
    if single_win(chr(c1), board):
        single_count += 1
    for c2 in range(ord('Z'), c1, -1):
        # 双牛获胜，team_win中已实现ZA和AZ为相同组合
        if team_win(chr(c1), chr(c2), board):
            team_count += 1

fout.write(f"{single_count}\n{team_count}\n")
fout.close()



